### PR TITLE
LFM2/LFM2.5 paged serving with prefix caching

### PR DESCRIPTION
**Pull request by WindChimeRan**
## Summary

(I've been working on this for a very long time, because I wanna settle the shared state management & align mode prefix caching from Qwen3.5+ first. Now they are all in a usable shape, so I pick it up. LFM is a good model for mac inference! )

Adds paged text generation for LFM2 and LFM2.5 dense and MoE models, with prefix caching. 

* **supported**: Prefix caching defaults to `align`(same with upstream vllm). 
* **unsupported**: `all` mode and speculative decoding remain unsupported.

* commit 1 is the full LFM support (conv layer, prefix caching .etc)
* commit 2 is pure renaming, because now we have two kinds of state now. I checked the upstream vllm naming, and use state for shared worker machinery, and retain GDN / ShortConv for the concrete implementations.

## Performance

Serving output throughput is generated output tokens divided by total benchmark wall time, including prefill and scheduling. These runs do not isolate decode-kernel speed.

Sonnet, nominal **1024 input + 128 output tokens**, **100 requests/run**, rate **10/s**, concurrency **32**. The figure shows medians of two fresh-server runs on an M5 Pro (64 GiB); all 1,600 requests succeeded. Memory fraction: 0.3 for the small pair, 0.5 for the 8B pair.

benchmark_comparison

<details>
<summary>Reproduce the performance comparison</summary>

Run from revision `761e583`, with the vllm-metal environment active. Tested with macOS 26.6.2, vLLM 0.28.0+cpu, MLX 0.32.0, MLX-LM `254d153`, mlx-vlm 0.6.4, and Transformers 5.12.1. Native checkpoint dtypes are preserved.

Select one row below, then run with `APC=off` and `APC=on`, twice each. Restart the server for every run; reverse model order in the second repetition. No client warmup or discarded measured runs.

| `MODEL` | `REVISION` | `FRACTION` |
|---|---|---:|
| `Qwen/Qwen3.5-0.8B` | `2fc06364715b967f1860aea9cf38778875588b17` | 0.3 |
| `LiquidAI/LFM2.5-1.2B-Instruct` | `df58c174f05ff733f83f8cae10ea9298224c8006` | 0.3 |
| `Qwen/Qwen3-8B` | `b968826d9c46dd6066d109eabc6255188de91218` | 0.5 |
| `LiquidAI/LFM2.5-8B-A1B` | `b9aebfcbe28b6cb374042f495d733037550ab146` | 0.5 |

```bash
# Bash; use a free port. Change these values for each model/mode/repetition.
set -eo pipefail
MODEL=LiquidAI/LFM2.5-1.2B-Instruct
REVISION=df58c174f05ff733f83f8cae10ea9298224c8006
FRACTION=0.3
APC=on
REP=1
PORT=8000

# Untimed setup: cache the pinned checkpoint and dataset.
HF_HUB_OFFLINE=0 hf download "$MODEL" --revision "$REVISION" \
  --include '*.safetensors' --include '*.json' --include '*.txt' \
  --include '*.model' --include '*.jinja'
MODEL_PATH=$(python - "$MODEL" "$REVISION" <<'PYCODE'
import sys
from pathlib import Path
from huggingface_hub import hf_hub_download
print(Path(hf_hub_download(sys.argv[1], 'config.json',
    revision=sys.argv[2], local_files_only=True)).parent)
PYCODE
)
curl -fsSL https://raw.githubusercontent.com/vllm-project/vllm/v0.28.0/benchmarks/sonnet.txt -o sonnet.txt
printf '%s  sonnet.txt\n' d58663195ba6780da5f029b920c7ac00cad1e435ee1df7e03bf9ec2470f8dea4 | shasum -a 256 -c -

export PYTHONPATH="$PWD" HF_HUB_OFFLINE=1
export VLLM_METAL_USE_PAGED_ATTENTION=1 VLLM_METAL_MEMORY_FRACTION="$FRACTION"
export VLLM_ENABLE_V1_MULTIPROCESSING=1 GLOO_SOCKET_IFNAME=lo0 TOKENIZERS_PARALLELISM=false
python -m vllm_metal.metal.build
CACHE_ARGS=()
if [ "$APC" = off ]; then CACHE_ARGS+=(--no-enable-prefix-caching); fi
RUN="${MODEL##*/}-${APC}-r${REP}"

python -m vllm.entrypoints.openai.api_server \
  --model "$MODEL_PATH" --served-model-name "$MODEL" \
  --max-model-len 2048 --host 127.0.0.1 --port "$PORT" \
  --generation-config vllm "${CACHE_ARGS[@]}" > "$RUN.server.log" 2>&1 &
SERVER_PID=$!
trap 'kill "$SERVER_PID" 2>/dev/null || true; wait "$SERVER_PID" 2>/dev/null || true' EXIT
until curl -fsS "http://127.0.0.1:$PORT/health" >/dev/null 2>&1; do
  kill -0 "$SERVER_PID" 2>/dev/null || exit 1
  sleep 1
done

python -m vllm.entrypoints.cli.main bench serve \
  --backend vllm --base-url "http://127.0.0.1:$PORT" \
  --model "$MODEL" --tokenizer "$MODEL_PATH" --endpoint /v1/completions \
  --dataset-name sonnet --dataset-path sonnet.txt \
  --sonnet-input-len 1024 --sonnet-output-len 128 --sonnet-prefix-len 200 \
  --num-prompts 100 --request-rate 10 --max-concurrency 32 \
  --temperature 0 --ignore-eos --seed 0 --num-warmups 0 \
  --save-result --save-detailed --result-dir benchmark-results \
  --result-filename "$RUN.json" --label "$RUN" --disable-tqdm \
  | tee "$RUN.client.log"

kill "$SERVER_PID"
wait "$SERVER_PID" || true
trap - EXIT
```

Sonnet samples whole lines: mean inputs were 941 vs 1,014 tokens for the small pair, and 1,012 for both 8B models. The 8B comparison matches total weight size, not FLOPs. The 200-token prefix is shorter than Qwen3.5's 544-token state block, giving it zero cache hits versus 17–19% for the other models; APC-off is the cleaner model comparison.

</details>

vllm-metal vs vllm-cpu

(FP32 is faster than BF16 in vllm-cpu)

serving_cpu_vs_metal

### Prefix caching

Upstream vLLM `prefix_repetition`: **1,152 shared + 128 unique input tokens**, 128 output tokens. Same hardware, models, memory fractions, 100 requests/run, rate 10/s, and concurrency 32 as above. Medians of two fresh-server runs, each seeded with one untimed request; all 1,600 measured requests succeeded.

apc_comparison

| Model | Serving output tok/s, off → on | Serving speedup | Mean TTFT (ms), off → on | Cached input tokens |
|---|---:|---:|---:|---:|
| Qwen3.5-0.8B | 474.5 → 684.1 | 1.44× | 824.4 → 143.7 | 85.0% |
| LFM2.5-1.2B | 644.9 → 1035.4 | 1.61× | 768.1 → 113.3 | 89.8% |
| Qwen3-8B | 84.4 → 166.1 | 1.97× | 6930.6 → 1177.8 | 89.9% |
| LFM2.5-8B-A1B | 211.9 → 279.2 | 1.32× | 1868.5 → 609.6 | 89.9% |

<details>
<summary>Reproduce the APC comparison</summary>

Use the same environment and pinned checkpoint table above. Run each model with `APC=off` and `APC=on`, twice each, restarting the server each time and reversing order in the second repetition. Set `FRACTION=0.5` for either 8B model. Native artifacts must be built as in the earlier reproduction. The measured runtime is `a64066f`; subsequent commits trim tests and rename shared state interfaces without changing the serving algorithms.

```bash
#!/usr/bin/env bash
# Run from the PR checkout with its environment active and checkpoint cached.
set -eo pipefail
MODEL=${MODEL:-LiquidAI/LFM2.5-1.2B-Instruct}
REVISION=${REVISION:-df58c174f05ff733f83f8cae10ea9298224c8006}
FRACTION=${FRACTION:-0.3}
APC=${APC:-on}
REP=${REP:-1}
PORT=${PORT:-8000}
MODEL_PATH=$(python - "$MODEL" "$REVISION" <<'PY'
import sys
from pathlib import Path
from huggingface_hub import hf_hub_download
print(Path(hf_hub_download(sys.argv[1], 'config.json',
    revision=sys.argv[2], local_files_only=True)).parent)
PY
)
export PYTHONPATH="$PWD" HF_HUB_OFFLINE=1
export VLLM_METAL_USE_PAGED_ATTENTION=1 VLLM_METAL_MEMORY_FRACTION="$FRACTION"
export VLLM_ENABLE_V1_MULTIPROCESSING=1 GLOO_SOCKET_IFNAME=lo0 TOKENIZERS_PARALLELISM=false
CACHE_ARGS=()
if [ "$APC" = off ]; then CACHE_ARGS+=(--no-enable-prefix-caching); fi
RUN="${MODEL##*/}-prefix-repetition-${APC}-r${REP}"
python -m vllm.entrypoints.openai.api_server \
  --model "$MODEL_PATH" --served-model-name "$MODEL" \
  --max-model-len 2048 --host 127.0.0.1 --port "$PORT" \
  --generation-config vllm "${CACHE_ARGS[@]}" > "$RUN.server.log" 2>&1 &
SERVER_PID=$!
trap 'kill "$SERVER_PID" 2>/dev/null || true; wait "$SERVER_PID" 2>/dev/null || true' EXIT
until curl -fsS "http://127.0.0.1:$PORT/health" >/dev/null 2>&1; do
  kill -0 "$SERVER_PID" 2>/dev/null || exit 1
  sleep 1
done

# Untimed seed, identical in both APC arms. Match completion-side BOS handling.
python - "$MODEL_PATH" "$MODEL" "$PORT" <<'PY'
import sys, requests
from vllm.benchmarks.datasets.datasets import PrefixRepetitionRandomDataset
from vllm.tokenizers import get_tokenizer
tokenizer = get_tokenizer(sys.argv[1])
samples = PrefixRepetitionRandomDataset(random_seed=0).sample(
    tokenizer=tokenizer, num_requests=100, prefix_len=1152,
    suffix_len=128, num_prefixes=1, output_len=128)
# 1088 is divisible by both default state block sizes (16 and 544).
# One extra prompt token preserves that boundary when decoding starts.
prefix = tokenizer.encode(samples[0].prompt, add_special_tokens=True)[:1089]
response = requests.post(f'http://127.0.0.1:{sys.argv[3]}/v1/completions',
    json=dict(model=sys.argv[2], prompt=prefix, max_tokens=1,
              temperature=0, ignore_eos=True), timeout=180)
response.raise_for_status()
assert response.json()['usage']['prompt_tokens'] == 1089
assert response.json()['usage']['completion_tokens'] == 1
PY
curl -fsS "http://127.0.0.1:$PORT/metrics" > "$RUN.metrics-before.txt"
python -m vllm.entrypoints.cli.main bench serve \
  --backend vllm --base-url "http://127.0.0.1:$PORT" \
  --model "$MODEL" --tokenizer "$MODEL_PATH" --endpoint /v1/completions \
  --dataset-name prefix_repetition \
  --prefix-repetition-prefix-len 1152 --prefix-repetition-suffix-len 128 \
  --prefix-repetition-num-prefixes 1 --prefix-repetition-output-len 128 \
  --num-prompts 100 --request-rate 10 --max-concurrency 32 \
  --temperature 0 --ignore-eos --seed 0 --num-warmups 0 \
  --ready-check-timeout-sec 0 \
  --save-result --save-detailed --result-dir benchmark-results \
  --result-filename "$RUN.json" --label "$RUN" --disable-tqdm \
  | tee "$RUN.client.log"
curl -fsS "http://127.0.0.1:$PORT/metrics" > "$RUN.metrics-after.txt"
kill "$SERVER_PID"
wait "$SERVER_PID" || true
trap - EXIT
```

The seed uses 1,089 prompt tokens and one output token to preserve the common 1,088-token checkpoint. Metrics are sampled after the seed, so its work is excluded. Dense LFM adds BOS, giving 1,281 input tokens per measured request; the other models use 1,280. Default state blocks remain 544 tokens for Qwen3.5 and 16 for the others.

Four initial dense-LFM runs were retained as diagnostics and repeated after correcting BOS handling in the untimed seed. The table uses the two corrected repetitions. This synthetic workload measures high prefix reuse; it does not represent general application traffic.

</details>

## Correctness

Manual native-checkpoint validation for LFM2.5-1.2B-Instruct and LFM2-1.2B: **19/19 exact token comparisons per checkpoint against independently run MLX**, using matched prefill chunks. Each match is an identical 12-token continuation. The local diagnostic covers singleton/concurrent generation, APC off, align, and fine-prefix modes.

| Checkpoint | Exact matches |
|---|---:|
| LFM2.5-1.2B-Instruct | 19/19 |
| LFM2-1.2B | 19/19 |
| LFM2.5-8B-A1B | 16/19 |

The separate HTTP test matches 5/5 outputs; three concurrent branches each reuse 160 tokens. MoE's remaining differences arise at numerical boundaries; ShortConv-only full-model traces are bit-exact to MLX.

<details>
<summary>Reproduce native-checkpoint correctness checks</summary>

Checkpoint-specific E2E checks are local diagnostics, excluded from the repository test suite. From the same checkout/environment, fetch the pinned harness into a temporary directory and run it against both dense checkpoints:

```bash
set -eo pipefail
export PYTHONPATH="$PWD" VLLM_METAL_LFM2_E2E=1
E2E_DIR=$(mktemp -d "${TMPDIR:-/tmp}/lfm2-diagnostic.XXXXXX")
E2E_SCRIPT="$E2E_DIR/checkpoint_e2e.py"
curl -fsSL \
  https://raw.githubusercontent.com/WindChimeRan/vllm-metal/3a32dfd52f6488410d333a20c87d1ff9b3a9eede/tests/test_lfm2_prefix_caching_e2e.py \
  -o "$E2E_SCRIPT"
printf '%s  %s\n' \
  30f1ceb20140f0718e15fd19b8798a726251eaef56cd7cbc2e738a7bc602185b \
  "$E2E_SCRIPT" | shasum -a 256 -c -
python -m vllm_metal.metal.build

for SPEC in \
  'LiquidAI/LFM2.5-1.2B-Instruct df58c174f05ff733f83f8cae10ea9298224c8006' \
  'LiquidAI/LFM2-1.2B 40f3da0d0164913923aee9462c23077868b816a3'
do
  read -r MODEL REVISION <<< "$SPEC"
  HF_HUB_OFFLINE=0 hf download "$MODEL" --revision "$REVISION" \
    --include '*.safetensors' --include '*.json' --include '*.txt' \
    --include '*.model' --include '*.jinja'
  MODEL_PATH=$(python - "$MODEL" "$REVISION" <<'PYCODE'
import sys
from pathlib import Path
from huggingface_hub import hf_hub_download
print(Path(hf_hub_download(sys.argv[1], 'config.json',
    revision=sys.argv[2], local_files_only=True)).parent)
PYCODE
)
  HF_HUB_OFFLINE=1 VLLM_METAL_MEMORY_FRACTION=0.15 \
  LFM2_MODEL_PATH="$MODEL_PATH" \
  python -m pytest "$E2E_SCRIPT" -q -s \
    --basetemp="$(mktemp -d "${TMPDIR:-/tmp}/lfm2-correctness.XXXXXX")"
done
```

For the MoE diagnostic, use `LiquidAI/LFM2.5-8B-A1B`, revision `b9aebfcbe28b6cb374042f495d733037550ab146`, with memory fraction `0.8`. Its strict comparison reports the three mismatches above and exits nonzero; no alternate expected tokens are accepted.

The reference uses full prompt chunks. Canonical `mlx_lm.generate_step` reserves the final prompt token for a separate call; five additional raw-prompt comparisons matched 4/5 sequences for each dense checkpoint and 2/5 for MoE. The controlled FP32 MoE trace first diverges in shared paged SDPA, then later expert selections amplify the difference. These results do not claim universal greedy identity across kernels or execution schedules.

</details>

## Roadmap

Built on merged #668 and #696. The native cache declaration matches the attention/`ArraysCache` split proposed in #675. The stateless layer role and Mamba-2 support remain with the Nemotron work in #644.

## Validation

- GDN prefix-cache E2E passes; all five smoke outputs match the prerequisite baseline, including one existing golden mismatch.
- Full non-slow suite: 1,986 passed, 14 skipped; nine GGUF numerical failures also reproduce on unchanged main.

### CODE REVIEW

File: vllm_metal/platform.py [hunk 1]
```diff
@@ -513,14 +513,31 @@ def check_and_update_config(cls, vllm_config: "VllmConfig") -> None:
 
         if model_config is not None and model_config.is_hybrid:
```
Comment (LxYuan0420) [inline vllm_metal/platform.py:514]: Move this out of `platform.py`. The platform should not hardcode `lfm2` / `lfm2_moe` or assume every non-ShortConv hybrid is GDN. Cache dtype policy should come from the resolved hybrid family, so the next state family does not need another platform branch.

Suggested shape:

```python
if model_config is not None and model_config.is_hybrid:
    family_policy = hybrid_family_policy_for(model_config.hf_config)
    family_policy.validate_cache_config(model_config, cache_config)
```

File: vllm_metal/attention/impls/shortconv.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    ) -> None:
+        object.__setattr__(self, "_state_cache", state_cache)
+        object.__setattr__(self, "_cache_idx", cache_idx)
+
+    def __call__(
+        self,
+        x: mx.array,
+        mask: mx.array | None = None,
+        cache: nn.Module | None = None,
+        **kwargs: Any,
```
Comment (LxYuan0420) [inline vllm_metal/attention/impls/shortconv.py]: Nit: remove `**kwargs`. Current mlx-lm ShortConv only takes `x`, `mask`, and `cache`, and this wrapper does not forward extra args. Keeping unused kwargs makes the contract look broader than it is.

File: vllm_metal/v1/cache_policy.py [hunk 3]
```diff
@@ -666,7 +668,7 @@ def _adopt_hybrid_scheduler_group(
         layer_pool_ordinals: list[int] | None = None
         if self._runner.cache_config.mamba_cache_mode == "align":
             cache_idx_by_name = {
```
Comment (LxYuan0420) [inline vllm_metal/v1/cache_policy.py:670]: Nit: this is now shared by GDN and ShortConv. Please rename the nearby `linear-attention layers` wording to `hybrid state layers`.

**Review by LxYuan0420 — APPROVED** [review]
Great work!

**Review by ricky-chaoju — APPROVED** [review]

File: vllm_metal/platform.py (hunk 1 above)
Comment (WindChimeRan) [reply vllm_metal/platform.py:514]: I agree that we should move it out of `platform.py` and we should not vender this model specific hardcode config. 

I found your https://github.com/vllm-project/vllm-metal/pull/699 this is a clearer way to address that. (read the vllm config, then torch to mlx)

But that's a bigger refactoring PR than a small change, and it requires many change to the existing GDN files. 

I'll make a followup PR to address this soon.
